src/core/server/http/base_path_proxy_server.ts

```diff
@@ -0,0 +1,162 @@
+import { DevConfig } from '../dev';
+import { Logger } from '../logging';
+import { HttpConfig } from './http_config';
+import { createServer, getServerOptions } from './http_tools';
+
+const alphabet = 'abcdefghijklmnopqrztuvwxyz'.split('');
+
+export interface BasePathProxyServerOptions {
+  httpConfig: HttpConfig;
+  devConfig: DevConfig;
+  isKibanaPath: (path: string) => boolean;
+  blockUntil?: () => Promise<void>;
+}
+
+export class BasePathProxyServer {
+  private server?: Server;
+  private httpsAgent?: HttpsAgent;
+
+  get basePath() {
+    return this.options.httpConfig.basePath;
+  }
+
+  get targetPort() {
+    return this.options.devConfig.basePathProxyTargetPort;
+  }
+
+  constructor(
+    private readonly log: Logger,
+    private readonly options: BasePathProxyServerOptions
+  ) {
+    const ONE_GIGABYTE = 1024 * 1024 * 1024;
+    options.httpConfig.maxPayload = new ByteSizeValue(ONE_GIGABYTE);
+
+    if (!options.httpConfig.basePath) {
+      options.httpConfig.basePath = `/${sample(alphabet, 3).join('')}`;
+    }
+  }
+
+  public async start() {
+    const { httpConfig, devConfig, blockUntil, isKibanaPath } = this.options;
+
+    const options = getServerOptions(httpConfig);
+    const server = createServer(options);
+
+    await server.register({ plugin: require('h2o2-latest') });
+
+    if (httpConfig.ssl.enabled) {
+      const tlsOptions = options.tls as TlsOptions;
+      this.httpsAgent = new HttpsAgent({
+        ca: tlsOptions.ca,
+        cert: tlsOptions.cert,
+        key: tlsOptions.key,
+        passphrase: tlsOptions.passphrase,
+        rejectUnauthorized: false,
+      });
+    }
+
+    // Always redirect from root URL to the URL with basepath.
+    server.route({
+      handler: (request, responseToolkit) => {
+        return responseToolkit.redirect(httpConfig.basePath);
+      },
+      method: 'GET',
+      path: '/',
+    });
+
+    // Before we proxy request to a target port we may want to wait until some
+    // condition is met (e.g. until target listener is ready).
+    let routeOptions: RouteOptions | undefined;
+    if (blockUntil !== undefined) {
+      routeOptions = {
+        pre: [
+          async (request, responseToolkit) => {
+            await blockUntil();
+            return responseToolkit.continue;
+          },
+        ],
+      };
+    }
+
+    server.route({
+      handler: {
+        proxy: {
+          agent: this.httpsAgent,
+          host: server.info.host,
+          passThrough: true,
+          port: devConfig.basePathProxyTargetPort,
+          protocol: server.info.protocol,
+          xforward: true,
+        },
+      },
+      method: '*',
+      options: routeOptions,
+      path: `${httpConfig.basePath}/{kbnPath*}`,
+    });
+
+    // It may happen that basepath has changed, but user still uses the old one,
+    // so we can try to check if that's the case and just redirect user to the
+    // same URL, but with valid basepath.
+    server.route({
```

Comment:
Sorry it looks like I'm failed to add more clarity to this weird thing from master, so I'll be really grateful if you can come up with a better wording! :) See https://github.com/elastic/kibana/blob/f0cc432fafd6f7053f32d493aba8b576dcb20539/src/cli/cluster/base_path_proxy.js#L115-L133
It took me some time to really understand what's going on here, but I think I finally got it: it's a dev-mode only thing, when you run kibana with `yarn start` we generate random base path (e.g. `/foo`), then for some reason you modified source code and decided to restart Kibana. After restart we'll have a new base path (e.g. `/bar`), but if you had already opened some Kibana web page in the browser with `/foo` base path it would be a pain to manually rewrite base path. So this route will allow you to just refresh web page with the old base path (`/foo`) and be redirected to the same page with new base path (`/bar`) automagically (yeah, I really classify this as magic).

Reply:
Ahhh, very interesting. I appreciate the explanation, thank you! So it's from master. 🆗